File: python/dash_app/charts.py
"""Constructeurs de graphiques.

Règles appliquées systématiquement :
  * un seul axe de valeurs — jamais deux échelles Y sur un même graphique ;
  * marques fines, extrémités arrondies, écart de 2 px entre segments empilés ;
  * légende dès 2 séries, étiquettes directes sélectives (jamais toutes) ;
  * grille discrète, axes en encre atténuée ;
  * survol actif par défaut (curseur unifié sur les courbes).
"""

from __future__ import annotations

import plotly.graph_objects as go

from erp.config import compact_money
from erp.theme import SERIES, STATUS, SURFACE, apply_layout, sequential_scale, series_color

HOVER = "<b>%{x}</b><br>%{y:,.0f}<extra></extra>"


def _fmt_axis(fig: go.Figure, symbol: str | None) -> None:
    """Graduations compactes (12 k, 4,2 M) : un axe monétaire brut est illisible."""
    if symbol:
        fig.update_yaxes(tickformat="~s", ticksuffix=f" {symbol}", automargin=True)
    else:
        fig.update_yaxes(tickformat="~s", automargin=True)


def fr_period(period: str) -> str:
    """'2026-02' -> 'févr. 2026' — et force l'axe en catégories, pas en dates."""
    mois = ["janv.", "févr.", "mars", "avr.", "mai", "juin",
            "juil.", "août", "sept.", "oct.", "nov.", "déc."]
    try:
        year, month = period.split("-")[:2]
        return f"{mois[int(month) - 1]} {year}"
    except (ValueError, IndexError):
        return period


def fr_day(day: str) -> str:
    """'2026-08-16' -> '16/08'."""
    try:
        y, m, d = day.split("-")
        return f"{d}/{m}"
    except ValueError:
        return day


def line_chart(x: list, series: dict[str, list], symbol: str | None = None,
               height: int = 290, fill: bool = False, title: str = "") -> go.Figure:
    """Évolution dans le temps. Courbes de 2 px, marqueurs masqués, survol unifié."""
    fig = go.Figure()
    for i, (label, values) in enumerate(series.items()):
        color = series_color(i)
        fig.add_trace(go.Scatter(
            x=x, y=values, name=label, mode="lines",
            line=dict(color=color, width=2, shape="spline", smoothing=0.4),
            fill="tozeroy" if (fill and len(series) == 1) else None,
            fillcolor=_alpha(color, 0.10) if fill else None,
            hovertemplate=f"{label} : %{{y:,.0f}}<extra></extra>",
        ))
    apply_layout(fig, title, height, showlegend=len(series) > 1)
    fig.update_xaxes(type="category", nticks=9, automargin=True)
    _fmt_axis(fig, symbol)
    return fig


def bar_chart(labels: list[str], values: list[float], symbol: str | None = None,
              horizontal: bool = True, height: int = 300, color: str | None = None,
              title: str = "", show_values: bool = True) -> go.Figure:
    """Comparaison de grandeurs. Une seule série -> une seule couleur, pas de légende."""
    c = color or series_color(0)
    text = [compact_money(v, symbol) if symbol else f"{v:,.0f}".replace(",", " ")
            for v in values] if show_values else None
    fig = go.Figure(go.Bar(
        x=values if horizontal else labels,
        y=labels if horizontal else values,
        orientation="h" if horizontal else "v",
        marker=dict(color=c, cornerradius=4),
        text=text, textposition="auto",
        textfont=dict(size=11.5, color=SURFACE["ink_secondary"]),
        hovertemplate="%{y}<br><b>%{x:,.0f}</b><extra></extra>" if horizontal
                      else "%{x}<br><b>%{y:,.0f}</b><extra></extra>",
    ))
    apply_layout(fig, title, height)
    if horizontal:
        fig.update_yaxes(autorange="reversed", showgrid=False, automargin=True)
        fig.update_xaxes(gridcolor=SURFACE["grid"], showgrid=True, showticklabels=False)
    else:
        _fmt_axis(fig, symbol)
    return fig


def grouped_bars(x: list[str], series: dict[str, list], symbol: str | None = None,
                 height: int = 300, title: str = "", stacked: bool = False) -> go.Figure:
    """Plusieurs séries comparables sur la MÊME échelle (sinon : deux graphiques)."""
    fig = go.Figure()
    for i, (label, values) in enumerate(series.items()):
        fig.add_trace(go.Bar(
            x=x, y=values, name=label,
            marker=dict(color=series_color(i), cornerradius=4,
                        line=dict(width=2, color=SURFACE["chart"]) if stacked else None),
            hovertemplate=f"{label} : %{{y:,.0f}}<extra></extra>",
        ))
    apply_layout(fig, title, height, showlegend=True,
                 barmode="stack" if stacked else "group", bargap=0.28, bargroupgap=0.08)
    fig.update_xaxes(type="category", automargin=True)
    _fmt_axis(fig, symbol)
    return fig


def waterfall(labels: list[str], values: list[float], symbol: str,
              height: int = 320, title: str = "") -> go.Figure:
    """Passage du chiffre d'affaires au résultat (paire divergente bleu / rouge)."""
    measures = ["absolute"] + ["relative"] * (len(values) - 2) + ["total"]
    fig = go.Figure(go.Waterfall(
        orientation="v", measure=measures, x=labels, y=values,
        text=[compact_money(v, symbol) for v in values],
        textposition="outside", textfont=dict(size=11.5, color=SURFACE["ink_secondary"]),
        connector=dict(line=dict(color=SURFACE["axis"], width=1)),
        increasing=dict(marker=dict(color=SERIES[0])),
        decreasing=dict(marker=dict(color=STATUS["critical"])),
        totals=dict(marker=dict(color=SERIES[6])),
        hovertemplate="%{x}<br><b>%{y:,.0f}</b><extra></extra>",
    ))
    apply_layout(fig, title, height)
    fig.update_xaxes(type="category", automargin=True)
    _fmt_axis(fig, symbol)
    return fig


def heatmap(z: list[list[float]], x: list[str], y: list[str], height: int = 300,
            title: str = "", unit: str = "") -> go.Figure:
    """Magnitude continue : rampe mono-teinte clair -> foncé, jamais un arc-en-ciel."""
    scale = sequential_scale(9)
    fig = go.Figure(go.Heatmap(
        z=z, x=x, y=y,
        colorscale=[[i / (len(scale) - 1), c] for i, c in enumerate(scale)],
        hovertemplate=f"%{{y}} · %{{x}}<br><b>%{{z:,.0f}}{unit}</b><extra></extra>",
        colorbar=dict(thickness=9, outlinewidth=0, len=0.85,
                      tickfont=dict(size=11, color=SURFACE["ink_muted"])),
        xgap=2, ygap=2,
    ))
    apply_layout(fig, title, height)
    fig.update_xaxes(showgrid=False)
    fig.update_yaxes(showgrid=False)
    return fig


def dumbbell(labels: list[str], left: list[float], right: list[float],
             left_name: str, right_name: str, symbol: str | None = None,
             height: int = 320, title: str = "") -> go.Figure:
    """Deux mesures de MÊME unité par catégorie (ex. valeur d'achat vs valeur de vente)."""
    fig = go.Figure()
    for i, lab in enumerate(labels):
        fig.add_trace(go.Scatter(
            x=[left[i], right[i]], y=[lab, lab], mode="lines",
            line=dict(color=SURFACE["axis"], width=2), hoverinfo="skip",
            showlegend=False,
        ))
    for values, name, idx in ((left, left_name, 0), (right, right_name, 2)):
        fig.add_trace(go.Scatter(
            x=values, y=labels, mode="markers", name=name,
            marker=dict(size=11, color=series_color(idx),
                        line=dict(width=2, color=SURFACE["chart"])),
            hovertemplate=f"{name} · %{{y}}<br><b>%{{x:,.0f}}</b><extra></extra>",
        ))
    apply_layout(fig, title, height, showlegend=True)
    fig.update_yaxes(autorange="reversed", showgrid=False, automargin=True)
    _fmt_axis(fig, symbol)
    return fig


def share_bars(labels: list[str], values: list[float], symbol: str,
               height: int = 300, title: str = "") -> go.Figure:
    """Répartition : barres horizontales triées + part en %, plus lisibles qu'un camembert."""
    total = sum(values) or 1
    text = [f"{compact_money(v, symbol)}  ·  {v / total * 100:.0f} %" for v in values]
    fig = go.Figure(go.Bar(
        x=values, y=labels, orientation="h",
        marker=dict(color=[series_color(i) for i in range(len(labels))], cornerradius=4),
        text=text, textposition="outside", cliponaxis=False,
        textfont=dict(size=11.5, color=SURFACE["ink_secondary"]),
        hovertemplate="%{y}<br><b>%{x:,.0f}</b><extra></extra>",
    ))
    apply_layout(fig, title, height)
    fig.update_yaxes(autorange="reversed", showgrid=False, automargin=True)
    fig.update_xaxes(showticklabels=False, showgrid=False,
                     range=[0, max(values) * 1.95 if values else 1])
    return fig


def _alpha(hex_color: str, alpha: float) -> str:
    h = hex_color.lstrip("#")
    r, g, b = (int(h[i:i + 2], 16) for i in (0, 2, 4))
    return f"rgba({r},{g},{b},{alpha})"

File: shiny/R/charts.R
# =====================================================================
#  Constructeurs de graphiques (plotly).
#
#  Règles appliquées partout, comme dans les autres interfaces :
#    * un seul axe de valeurs, jamais deux échelles Y ;
#    * marques fines, extrémités arrondies, 2 px entre segments empilés ;
#    * légende dès deux séries, étiquettes directes sélectives ;
#    * grille discrète, axes en encre atténuée, survol actif.
# =====================================================================

#' Mise en forme commune (grille, axes, police, marges).
apply_layout <- function(p, height = 290, showlegend = FALSE, symbol = NULL, ...) {
  p <- plotly::layout(
    p,
    height = height,
    showlegend = showlegend,
    font = list(family = FONT, size = 13, color = SURFACE$ink_secondary),
    paper_bgcolor = SURFACE$chart,
    plot_bgcolor  = SURFACE$chart,
    margin = list(l = 60, r = 18, t = 18, b = 44),
    hovermode = "x unified",
    hoverlabel = list(font = list(family = FONT, size = 12), bgcolor = "#ffffff"),
    legend = list(orientation = "h", y = 1.08, x = 0,
                  font = list(size = 12, color = SURFACE$ink_secondary)),
    xaxis = list(showgrid = FALSE, zeroline = FALSE, linecolor = SURFACE$axis,
                 tickfont = list(color = SURFACE$ink_muted, size = 11),
                 type = "category", automargin = TRUE),
    yaxis = list(gridcolor = SURFACE$grid, zeroline = FALSE, showline = FALSE,
                 tickformat = "~s",
                 ticksuffix = if (is.null(symbol)) "" else paste0(" ", symbol),
                 tickfont = list(color = SURFACE$ink_muted, size = 11),
                 automargin = TRUE),
    ...
  )
  plotly::config(p, displayModeBar = FALSE, locale = "fr")
}

#' Courbe d'évolution. `series` est une liste nommée de vecteurs.
chart_line <- function(x, series, symbol = NULL, height = 280, fill = FALSE) {
  if (length(x) == 0) return(chart_empty("Aucune donnée sur la période."))
  p <- plotly::plot_ly()
  names_series <- names(series)
  for (i in seq_along(series)) {
    color <- series_color(i)
    p <- plotly::add_trace(
      p, x = x, y = series[[i]], name = names_series[i],
      type = "scatter", mode = "lines",
      line = list(color = color, width = 2, shape = "spline", smoothing = 0.4),
      fill = if (fill && length(series) == 1) "tozeroy" else "none",
      fillcolor = if (fill) paste0(color, "1A") else NULL,
      hovertemplate = paste0(names_series[i], " : %{y:,.0f}<extra></extra>"))
  }
  apply_layout(p, height, showlegend = length(series) > 1, symbol = symbol)
}

#' Barres verticales, une série, une couleur, pas de légende.
chart_bars <- function(labels, values, symbol = NULL, height = 280, color = NULL) {
  if (length(labels) == 0) return(chart_empty("Aucune donnée."))
  p <- plotly::plot_ly(
    x = labels, y = values, type = "bar",
    marker = list(color = color %||% series_color(1), cornerradius = 4),
    hovertemplate = "%{x}<br><b>%{y:,.0f}</b><extra></extra>")
  apply_layout(p, height, symbol = symbol, bargap = 0.3)
}

#' Barres groupées ou empilées, séries de MÊME unité.
chart_grouped <- function(x, series, symbol = NULL, height = 300, stacked = FALSE) {
  if (length(x) == 0) return(chart_empty("Aucune donnée."))
  p <- plotly::plot_ly()
  names_series <- names(series)
  for (i in seq_along(series)) {
    p <- plotly::add_trace(
      p, x = x, y = series[[i]], name = names_series[i], type = "bar",
      marker = list(color = series_color(i), cornerradius = 4,
                    line = if (stacked) list(width = 2, color = SURFACE$chart) else NULL),
      hovertemplate = paste0(names_series[i], " : %{y:,.0f}<extra></extra>"))
  }
  apply_layout(p, height, showlegend = TRUE, symbol = symbol,
               barmode = if (stacked) "stack" else "group",
               bargap = 0.28, bargroupgap = 0.08)
}

#' Répartition : barres horizontales triées, valeur et part affichées.
chart_share <- function(labels, values, symbol = "", height = NULL) {
  if (length(labels) == 0) return(chart_empty("Aucune donnée."))
  height <- height %||% max(200, 46 * length(labels) + 60)
  total <- sum(values)
  if (total == 0) total <- 1
  text <- sprintf("%s  ·  %d %%", vapply(values, fmt_compact, "", symbol),
                  round(values / total * 100))
  p <- plotly::plot_ly(
    x = values, y = factor(labels, levels = rev(labels)), type = "bar", orientation = "h",
    marker = list(color = vapply(seq_along(labels), series_color, ""), cornerradius = 4),
    text = text, textposition = "outside", cliponaxis = FALSE,
    textfont = list(size = 11.5, color = SURFACE$ink_secondary),
    hovertemplate = "%{y}<br><b>%{x:,.0f}</b><extra></extra>")
  p <- apply_layout(p, height, symbol = NULL)
  plotly::layout(
    p, hovermode = "closest",
    xaxis = list(showticklabels = FALSE, showgrid = FALSE, zeroline = FALSE,
                 range = c(0, max(values) * 1.9)),
    yaxis = list(showgrid = FALSE, automargin = TRUE, type = "category",
                 tickfont = list(color = SURFACE$ink_secondary, size = 12)))
}

#' Haltères : deux mesures de MÊME unité par catégorie.
chart_dumbbell <- function(labels, left, right, left_name, right_name,
                           symbol = NULL, height = NULL) {
  if (length(labels) == 0) return(chart_empty("Aucune donnée."))
  height <- height %||% max(220, 48 * length(labels) + 70)
  ordered <- factor(labels, levels = rev(labels))
  p <- plotly::plot_ly()
  for (i in seq_along(labels)) {
    p <- plotly::add_trace(
      p, x = c(left[i], right[i]), y = c(ordered[i], ordered[i]),
      type = "scatter", mode = "lines", showlegend = FALSE, hoverinfo = "skip",
      line = list(color = SURFACE$axis, width = 2))
  }
  p <- plotly::add_trace(
    p, x = left, y = ordered, type = "scatter", mode = "markers", name = left_name,
    marker = list(size = 11, color = series_color(1),
                  line = list(width = 2, color = SURFACE$chart)),
    hovertemplate = paste0(left_name, " · %{y}<br><b>%{x:,.0f}</b><extra></extra>"))
  p <- plotly::add_trace(
    p, x = right, y = ordered, type = "scatter", mode = "markers", name = right_name,
    marker = list(size = 11, color = series_color(3),
                  line = list(width = 2, color = SURFACE$chart)),
    hovertemplate = paste0(right_name, " · %{y}<br><b>%{x:,.0f}</b><extra></extra>"))
  p <- apply_layout(p, height, showlegend = TRUE, symbol = NULL)
  plotly::layout(
    p, hovermode = "closest",
    xaxis = list(tickformat = "~s", gridcolor = SURFACE$grid, zeroline = FALSE,
                 tickfont = list(color = SURFACE$ink_muted, size = 11)),
    yaxis = list(showgrid = FALSE, automargin = TRUE, type = "category",
                 tickfont = list(color = SURFACE$ink_secondary, size = 12)))
}

#' Carte de chaleur : magnitude continue, rampe mono-teinte.
chart_heatmap <- function(z, x, y, height = 300, unit = "") {
  scale <- sequential_scale(9)
  colorscale <- lapply(seq_along(scale), function(i) {
    list((i - 1) / (length(scale) - 1), scale[i])
  })
  p <- plotly::plot_ly(
    z = z, x = x, y = y, type = "heatmap", colorscale = colorscale,
    xgap = 2, ygap = 2,
    colorbar = list(thickness = 9, outlinewidth = 0, len = 0.85,
                    tickfont = list(size = 11, color = SURFACE$ink_muted)),
    hovertemplate = paste0("%{y} · %{x}<br><b>%{z:,.0f}", unit, "</b><extra></extra>"))
  p <- apply_layout(p, height)
  plotly::layout(p, hovermode = "closest",
                 yaxis = list(showgrid = FALSE, automargin = TRUE, type = "category",
                              tickfont = list(color = SURFACE$ink_muted, size = 11)))
}

#' Graphique vide (message centré), évite une zone blanche sans explication.
chart_empty <- function(message) {
  plotly::config(
    plotly::layout(
      plotly::plot_ly(type = "scatter", mode = "markers", x = numeric(0), y = numeric(0)),
      height = 200, paper_bgcolor = SURFACE$chart, plot_bgcolor = SURFACE$chart,
      xaxis = list(visible = FALSE), yaxis = list(visible = FALSE),
      annotations = list(list(text = message, showarrow = FALSE, xref = "paper",
                              yref = "paper", x = 0.5, y = 0.5,
                              font = list(size = 13, color = SURFACE$ink_muted)))),
    displayModeBar = FALSE)
}
